p2p/src/sessions/mod.rs

```diff
@@ -0,0 +1,181 @@
+//! Library for managing the sessions
+use std::collections::HashMap;
+use std::net::SocketAddr;
+
+use crate::sessions::error::{SessionsError, SessionsErrorKind, SessionsResult};
+use witnet_util::error::WitnetError;
+
+pub mod error;
+
+type SessionsMap<T> = HashMap<SocketAddr, SessionInfo<T>>;
+
+/// Session type
+#[derive(Copy, Clone, Debug)]
+pub enum SessionType {
+    /// Inbound session
+    Inbound,
+    /// Outbound session
+    Outbound,
+}
+
+/// Session Status (used for bootstrapping)
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum SessionStatus {
+    /// Recently created session (no handshake yet)
+    Unconsolidated,
+    /// Session with successful handshake
+    Consolidated,
+}
+
+/// Session info
+pub struct SessionInfo<T> {
+    /// Session reference (e.g. actor address)
+    pub reference: T,
+    /// Session status
+    pub status: SessionStatus,
+}
+
+/// Sessions
+/// This struct contains:
+/// - server address used to listen to incoming connections
+/// - lists of of inbound and outbound sessions parametrized with their reference (T)
+/// - sessions limit to be connected
+pub struct Sessions<T> {
+    // Server address listening to incoming connections
+    pub server_address: Option<SocketAddr>,
+    /// Inbound sessions: __untrusted__ peers that connect to the server
+    pub inbound_sessions: SessionsMap<T>,
+    /// Outbound sessions: __known__ peers that the node is connected
+    pub outbound_sessions: SessionsMap<T>,
+    // Inbound sessions limit
+    pub inbound_limit: usize,
+    // Outbound sessions limit
+    pub outbound_limit: usize,
+}
+
+impl<T> Default for Sessions<T> {
```

Comment:
it seems `Default ` can be 'derive`d directly here.

Reply:
You are right, we already tried to derive the `Default` trait, but we got the following error:
```rust
error[E0277]: the trait bound `actors::codec::_impl_act_request::actix::Addr<actors::session::Session>: std::default::Default` is not satisfied
  --> core/src/actors/sessions_manager.rs:29:5
   |
29 |     sessions: Sessions<Addr<Session>>,
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `std::default::Default` is not implemented for `actors::codec::_impl_act_request::actix::Addr<actors::session::Session>`
   |
   = note: required because of the requirements on the impl of `std::default::Default` for `witnet_p2p::sessions::Sessions<actors::codec::_impl_act_request::actix::Addr<actors::session::Session>>`
   = note: required by `std::default::Default::default`
```

Perhaps in this case, because T is `Addr<Session>`? Any idea?

Reply:
oops, my fault. this cannot be derived, since trait's orphan rules prevents impl `Default` for RemoteItem here.